api/all/src/main/java/io/opentelemetry/api/common/ComplexAttribute.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.api.common;
+
+import static io.opentelemetry.api.common.ArrayBackedComplexAttribute.sortAndFilterToAttributes;
+
+import java.util.Map;
+import java.util.function.BiConsumer;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * An immutable container for a complex attribute.
+ *
+ * <p>The keys are {@link AttributeKey}s and the values are Object instances that match the type of
+ * the provided key.
+ *
+ * <p>Null keys will be silently dropped.
+ *
+ * <p>Note: The behavior of null-valued attributes is undefined, and hence strongly discouraged.
+ *
+ * <p>Implementations of this interface *must* be immutable and have well-defined value-based
+ * equals/hashCode implementations. If an implementation does not strictly conform to these
+ * requirements, behavior of the OpenTelemetry APIs and default SDK cannot be guaranteed.
+ *
+ * <p>For this reason, it is strongly suggested that you use the implementation that is provided
+ * here via the factory methods and the {@link ComplexAttributeBuilder}.
+ */
+@SuppressWarnings("rawtypes")
+@Immutable
+public interface ComplexAttribute {
```

Comment:
I think that since this is so similar to `Attributes` that this should be plural: `ComplexAttributes`.

Reply:
it seems more like a single complex attribute value to me, e.g.

```
Logger.setAttribute("xyz", complexAttribute)
```

as it's not really a collection of complex attributes, it's a collection of attributes (both standard and complex)

I could see the interface named something like `NestedAttributes` though...